### ISSUE TITLE
R-script

### ISSUE THREAD

**Issue by f-has** [f-has/project1#1]
################ Clean and Import Data #############################

#import data
raw_data<-read.csv("/Users/fatimashire/Library/Mobile Documents/com~apple~CloudDocs/Applied Machine Learning/Assignments/cleveland.txt",header = FALSE)

#rename columns
names(raw_data)<-c("age","sex","cp","trestbps","chol","fbs","restecg","thalach","exang","oldpeak","slope","ca","thal","num")

#rename "?" as NA.
raw_data[raw_data=="?"]<-NA

#check type of datframes in r 
sapply(raw_data,typeof)

#group num factor if greater than 0 to 1, to reduce levels. Number 1 indicates presence of heart disease
raw_data$num[raw_data$num>0]<-1

#change types to numeric or factor based on continuous or categorical.
#change age,trestbps,chol,thalach,oldpeak to numeric
raw_data[,c(1,4,5,8,10)]<-lapply(raw_data[,c(1,4,5,8,10)],as.numeric)

#change sex, co(chest pain), fbs,restecg,exang,slope, ca,thal,num, to numeric
raw_data[,c(2,3,6,7,9,11:14)]<-lapply(raw_data[,c(2,3,6,7,9,11:14)],as.factor)

#rename factors for sex, cp(chest pain), fbs,restecg,exang,slope, ca,thal.
levels(raw_data$sex) <- c("female", "male")
levels(raw_data$cp) <- c("typical","atypical","non-anginal","asymptomatic")
levels(raw_data$fbs) <- c("false", "true")
levels(raw_data$restecg) <- c("normal","stt","hypertrophy")
levels(raw_data$exang) <- c("no","yes")
levels(raw_data$slope) <- c("upsloping","flat","downsloping")
levels(raw_data$thal) <- c("normal","fixed","reversable")
levels(raw_data$num) <- c("no heart disease","heart disease")

#remove NA rows
heart_data.nonmis<-na.omit(raw_data)
summary(heart_data.nonmis)

#Standardize data so we can compare models.

heart_data.nonmis <- scale(heart_data.nonmis)

library(glmnet)#load glmnet package
library(epiR)#load epiR package

#laod arm package
library(arm)

                                 #######plot all variables and visualize ############
par(mfrow=c(3,2))
boxplot(heart_data.nonmis$age~heart_data.nonmis$num, ylab="age(years)",xlab="num(units)")
boxplot(heart_data.nonmis$oldpeak~heart_data.nonmis$num,ylab = "oldpeak(units)",xlab="num(units)")
boxplot(heart_data.nonmis$trestbps~heart_data.nonmis$num,ylab="trestbps(mm Hg)",xlab="num(units)")
boxplot(heart_data.nonmis$chol~heart_data.nonmis$num, ylab="chol (mg/dl)",xlab="num(units)")
boxplot(heart_data.nonmis$thalach~heart_data.nonmis$num,ylab="thalach(units)",xlab="num(units)")

par(mfrow=c(3,3))
discrete.hist(heart_data.nonmis$cp,xlab="cp(units)")
discrete.hist(heart_data.nonmis$restecg,xlab="restec(units)")
discrete.hist(heart_data.nonmis$fbs,xlab="fbs(units)")
discrete.hist(heart_data.nonmis$exang,xlab="exang(units)")
discrete.hist(heart_data.nonmis$sex, xlab="sex(units)")
discrete.hist(heart_data.nonmis$ca,xlab="ca(units)")
discrete.hist(heart_data.nonmis$thal,xlab="thal(units)")
discrete.hist(heart_data.nonmis$num,xlab="num(units)")
discrete.hist(heart_data.nonmis$slope,xlab="slope(units)")
#calculate summary statistic:

#group into dataframe
sum.stats<-data.frame(unclass(summary(heart_data.nonmis)), check.names = FALSE, stringsAsFactors = FALSE)

#subset df into categorical and continous variables
cat_sum<-sum.stats[,-c(1,4,5,8,10)]
cont_sum<-sum.stats[,c(1,4,5,8,10)]

#export stats tables:
write.csv(cat_sum,"/Users/fatimashire/Library/Mobile Documents/com~apple~CloudDocs/Applied Machine Learning/Assignments/cat.sum.csv")
write.csv(cont_sum,"/Users/fatimashire/Library/Mobile Documents/com~apple~CloudDocs/Applied Machine Learning/Assignments/cont.sum.csv")

                            ####Apply lasso regression  #####

set.seed(12345)
sens.vec.l<-c() #empty vector for sensitivity
spec.vec.l<-c() #empty vector for specifcity
par(mfrow=c(3,2))#Set up a 5 x 2 plotting space
cv.l<-c()
for (i in 1:5){
train.I <- sample(nrow(heart_data.nonmis),round(nrow(heart_data.nonmis)*2/3))#split data into test and train
train.data <- heart_data.nonmis[train.I,]#split data into test and train
test.data <- heart_data.nonmis[-train.I,]#split data into test and train

# set x to matrix of predictors
x.train <- model.matrix(num ~.,train.data)[,-1]
#set vector of response
y.train<-train.data$num 

#fit the lasso model
l.mod <- glmnet(x.train,y.train,family="binomial", alpha=1)#generate lassmodel
#plot model
plot(l.mod,xvar = "lambda")#plot Log lambda against. 
legend("topright", lwd = 1, col=1:12,legend = colnames(x.train), cex = .4)#legend for predictors

# 10 fold crossvalidation on training model using alpha=1 
cv.l<- cv.glmnet(x.train,y.train,family="binomial", alpha=1)#predict linear model using lasso 

#plot(cv.l)
plot(cv.l)

#find optimal value to predict lambda
cv.l$lambda.min

#find optimal value to predict
coef.min.l <- coef(cv.l, s = "lambda.min")#determine min coefficient
coef.min.l#retrieve minimum coefficient 

#narrow down model selection,select coefficients that are not zero
rownames(coef.min.l)[coef.min.l[,1]!=0][-1]

#apply model on test data to make predictions
x.test <- model.matrix(num ~.,test.data)[,-1]# get predictor matrix

y.test<-test.data$num# set vector of response for test

predicted.l<-predict(cv.l, newx = x.test,s=cv.l$lambda.min ,type = "class")# predict model using lambda min using test data

#test performance of model using confusion matrix- predicted test data against true value for test data

conf.l<-table(predicted.l,y.test) [,2:1]#generate confusion matrix
epi.table<-epi.tests(conf.l)#generate the epitest to measure sensitivity and specificity
sens.vec.l[i]<-epi.table[["detail"]][["se"]][["est"]]#retrieve sensitivity vector
spec.vec.l[i]<-epi.table[["detail"]][["sp"]][["est"]]#retrieve sensitivity vector

}

acc.l.test.df<-data.frame("Sensitivity"=sens.vec.l,"Specificity"=spec.vec.l,average=(sens.vec.l+spec.vec.l)/2)#export performance measure vector for lasso model

#export sensitivity and specificty from testing set:
write.csv(acc.l.test.df,"/Users/fatimashire/Library/Mobile Documents/com~apple~CloudDocs/Applied Machine Learning/Assignments/lasso.test.perf.csv")

#generate average of spensitivity and specificity 
mean(sens.vec.l) #75% sensitivity
mean(spec.vec.l)#89% specificity

                                #######apply KNN:###### 

set.seed(12345)          

library(class)#load class package for knn cv

  missclass.err<-c()
  sens.knn.test.vec<-c()#sens vector used for to store sensitivity
  spec.knn.test.vec<-c()#sens vector used to store specificity
  par(mfrow=c(3,2))#Set up a 3 x 2 plotting space
  
  for(i in 1:5){
    
    for (j in 1:40){
      train.I <- sample(nrow(heart_data.nonmis),round(nrow(heart_data.nonmis)*2/3))
      train.data <- heart_data.nonmis[train.I,]
      test.data <- heart_data.nonmis[-train.I,]
    # set x train to matrix of predictors
    x.train <- model.matrix(num ~.,train.data)[,-1]
    #set y trainvector
    y.train<-train.data$num
    # set x test to matrix of predictors
    x.test <- model.matrix(num ~.,test.data)[,-1]
    #set response test vector
    y.test<-test.data$num
    predicted.knn.train <- knn.cv(x.train,y.train,k=j)#apply knn cross validations using training data, considering k=1:40 to find optimal k
    missclass.err[j]<-mean(predicted.knn.train!=y.train)}#calculate missclassification error
    knn.df<-data.frame("k"=1:40,"error"=missclass.err)#generate df to store missclass error and k value
  #test model with k max value from knn for loop 5 times to get sensitivity and specifcity for 5 trials:
  predicted.knn.test <- knn.cv(x.test,y.test,k=which.min(missclass.err))#apply knn cross validations using training data using LOO.
  conf.knn.test<-table(predicted.knn.test,test.data$num)[2:1,2:1]#generate confusion matrix
  epi.table.test<-epi.tests(conf.knn.test)#generate the epitest to measure sensitivity and specificity
  sens.knn.test.vec[i]<-epi.table.test[["detail"]][["se"]][["est"]]#retrieve sensitivity vector for train set
  spec.knn.test.vec[i]<-epi.table.test[["detail"]][["sp"]][["est"]]#retrieve sensitivity vector for train set
  #plot missclassification error against knn
  plot(knn.df$k,knn.df$error,xlab="K", ylab="Missclassification Error Rate (Units)",type="l")
}

#generate dataframe of sensitivity and specificty from test set.
acc.knn.test.df<-data.frame("Sensitivity"=sens.knn.test.vec,"Specificity"=spec.knn.test.vec,average=(spec.knn.test.vec+sens.knn.test.vec)/2)

#generate average of spensitivity and specificity 
mean(sens.knn.test.vec)  #39.4% sensitivity
mean( spec.knn.test.vec) #72.8%specificity

#export sensitivity and specificty from training set:
write.csv(acc.knn.train.df,"/Users/fatimashire/Library/Mobile Documents/com~apple~CloudDocs/Applied Machine Learning/Assignments/knn.train.perf.csv")
#export sensitivity and specificty from testing set:
write.csv(acc.knn.test.df,"/Users/fatimashire/Library/Mobile Documents/com~apple~CloudDocs/Applied Machine Learning/Assignments/knn.test.perf.csv")#export performance measure vector for knn model

                    ###Apply trees classification###
#load packages tree
#load package rpart
par(mfrow=c(3,2))#Set up a 3 x 2 plotting space
set.seed(12345)
library(rpart)
library(tree)
sens.vec.tree<-c()#sens vector for assesing performance
spec.vec.tree<-c()#spec vector for assesing performance
vec.sizes<-c()
for(i in 1:5){
  train.I <- sample(nrow(heart_data.nonmis),round(nrow(heart_data.nonmis)*2/3))#train random sample where to split data
  train.data <- heart_data.nonmis[train.I,]#create trained df
  test.data <- heart_data.nonmis[-train.I,]#create test df
  tree.mod<-tree(num~age+sex+cp+trestbps+chol+fbs+restecg+thalach+exang+oldpeak+slope+ca+thal,data=train.data)#apply tree model 
  
  cv.heart <- cv.tree(tree.mod, FUN=prune.tree, method = "misclass")#prune the tree using training data
  plot(cv.heart$size, cv.heart$dev / nrow(train.data), type = "b",
      xlab = "Tree Size", ylab = "Error Rate(units)")# plot missclassifcation error rate, pick number 6 based on plot

   # prune tree plot prunned classification tree using minimum deviance and prune at that size
  pruned.tree <- prune.misclass(tree.mod,best= cv.heart$size[[which.min(cv.heart$dev)]])
  plot(pruned.tree)#plot pruned tree
  text(pruned.tree,pretty=0,cex=0.5)#add text
  title(main = "Pruned Classification Tree")#add title
  
  #assess performance using pruned tree on test data: 
predicted.tree<-predict(pruned.tree,test.data,type="class")
conf.tree<-table(predicted=predicted.tree,actual=test.data$num) [2:1,2:1]#generate confusion matrix
epi.table<-epi.tests(conf.tree)#generate the epitest to measure sensitivity and specificity
sens.vec.tree[i]<-epi.table[["detail"]][["se"]][["est"]]#retrieve sensitivity vector
spec.vec.tree[i]<-epi.table[["detail"]][["sp"]][["est"]]#retrieve sensitivity vector

}
                      
#generate average of spensitivity and specificity 
mean(sens.vec.tree) # average sensitivity 73%
mean(spec.vec.tree)# average specificity 81%

#generate dataframe of sensitivity and specificty from test set.
acc.tree.test.df<-data.frame("Sensitivity"=sens.vec.tree,"Specificity"=spec.vec.tree,"average"=(spec.vec.tree+sens.vec.tree)/2)

write.csv(acc.tree.test.df,"/Users/fatimashire/Library/Mobile Documents/com~apple~CloudDocs/Applied Machine Learning/Assignments/tree.perf.csv")#export performance measure vector for knn model

                      ##############Train a model that can be used for classification of future observations#################
                                                        ################# Will use lasso ########
set.seed(12345)
#use entire data set.
# set x to matrix of predictors
x <- model.matrix(num ~.,heart_data.nonmis)[,-1]
#set y variable as num
y<-heart_data.nonmis$num

#fit the lasso model
l.mod.final <- glmnet(x,y,family="binomial", alpha=1)#generate final lassomodel
#plot model
plot(l.mod.final,xvar = "lambda", label = TRUE)#plot Log lambda against. 
legend("topright", lwd = 1, col=1:12,legend = colnames(x), cex = .4)#legend for predictors

# 10 fold cross-validation on training the  model using alpha=1 
cv.l.mod.final<- cv.glmnet(x,y,family="binomial", alpha=1)#predict linear model using lasso 

#plot(cv.l.mod.final)
plot(cv.l.mod.final)

#find optimal value to predict lambda
cv.l.mod.final$lambda.min #optimal lambda value is 0.007367976

#find optimal value to predict
coef.min.l.final <- coef(cv.l.mod.final, s = "lambda.min")#determine the min coefficient
coef.min.l.final#retrieve minimum coefficient 

#narrow down model selection,select coefficients that are not zero
rownames(coef.min.l.final)[coef.min.l.final[,1]!=0][-1]
